File: xiaoming/1011/PalindromeLinkedList.java

public class PalindromeLinkedList {
	
	public boolean isPalindrome(ListNode head) {
        
        if(head == null || head.next == null)
            return true;
        ListNode fast = head;
        ListNode slow = head;
        while(fast.next!=null && fast.next.next!=null){
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode secondHead = slow.next;
        slow.next = null;
        ListNode next = secondHead.next;
        secondHead.next = null;
     
        while(next != null){
            ListNode temp = next.next;
            next.next = secondHead;
            secondHead = next;
            next = temp;
        }
    
        while(head != null){
            if(head.val != secondHead.val)
                return false;
            head = head.next;
            secondHead = secondHead.next;
        }
        return true;
    }

}

File: mu_wang/10_11/Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        value = list()
        while head:
            value.append(head.val)
            head = head.next
            
        return [ x for x in value] == [ x for x in reversed(value)]